File: javalin/src/test/java/io/javalin/examples/HelloWorldAsync.java
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.examples;

import io.javalin.Javalin;

public class HelloWorldAsync {

    public static void main(String[] args) {
        Javalin app = Javalin.create(config -> {
            config.routes.get("/", ctx -> {
                ctx.async(
                    (asyncConfig) -> {
                        asyncConfig.timeout = 1000L;
                        asyncConfig.onTimeout(timeoutCtx -> timeoutCtx.result("Request timed out :<"));
                    },
                    () -> {
                        Thread.sleep((long) (Math.random() * 2000L));
                        ctx.result("Hello Javalin");
                    }
                );
            });
        }).start(7070);
    }

}

File: javalin/src/test/kotlin/io/javalin/examples/HelloWorldAsync.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.examples

import io.javalin.Javalin
import io.javalin.apibuilder.ApiBuilder.get
import io.javalin.plugin.bundled.DevLoggingPlugin
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

fun main() {
    Javalin.create {
        val scheduledExecutor = Executors.newSingleThreadScheduledExecutor()

        it.routes.apiBuilder {
            get("/result") { ctx ->
                ctx.future {
                    val future = CompletableFuture<String>()
                    scheduledExecutor.schedule({ future.complete("Hello World!") }, 10, TimeUnit.MILLISECONDS)
                    future.thenApply { ctx.result(it) }
                }
            }
            get("/json") { ctx ->
                ctx.future {
                    val future = CompletableFuture<List<String>>()
                    scheduledExecutor.schedule({ future.complete(listOf("a", "b", "c")) }, 10, TimeUnit.MILLISECONDS)
                    future.thenApply { ctx.json(it) }
                }
            }
        }

        it.registerPlugin(DevLoggingPlugin())
    }.start(7070)
}
